### PR TITLE
Refactor the Transition enum, collapsing enum variants that were

### PR DESCRIPTION
**Pull request by dabreegster**
combining primitive transitions into sequences.

Brief context on the state/transition system: The game crate is
organized as a stack of states, with the topmost one being active.
Transitions manipulate this stack. For example, the stack might look
like: [main menu, sandbox mode, edit mode, traffic signal editor, signal
picker]

@michaelkirk

### CODE REVIEW

File: game/src/edit/mod.rs [hunk 1]
```diff
@@ -97,10 +97,11 @@ impl EditMode {
             app.primary.clear_sim();
             if app.opts.resume_after_edit {
                 if self.mode.reset_after_edits() {
-                    Transition::PopThenReplaceThenPush(
-                        Box::new(SandboxMode::new(ctx, app, self.mode.clone())),
-                        TimeWarpScreen::new(ctx, app, old_sim.time(), false),
-                    )
+                    Transition::Multi(vec![
```
Comment (michaelkirk) [inline game/src/edit/mod.rs:100]: much better!

File: game/src/info/mod.rs [hunk 2]
```diff
@@ -528,14 +528,15 @@ impl InfoPanel {
...
-                            move |state, ctx, app| {
-                                let mut sandbox = state.downcast::<SandboxMode>().ok().unwrap();
-
+                            move |mut state, ctx, app| {
                                 if time < app.primary.sim.time() {
-                                    sandbox = ctx.loading_screen("rewind simulation", |ctx, _| {
-                                        Box::new(SandboxMode::new(ctx, app, sandbox.gameplay_mode))
+                                    let gameplay =
+                                        state.downcast::<SandboxMode>().ok().unwrap().gameplay_mode;
+                                    state = ctx.loading_screen("rewind simulation", |ctx, _| {
```
Comment (michaelkirk) [inline game/src/info/mod.rs:535]: I'm having a little trouble understanding the significance of this change - why are we mutating state in this new way, and getting gameplay differently?

File: game/src/game.rs [hunk 3]
```diff
@@ -77,83 +77,75 @@ impl Game {
...
-}
-
-impl GUI for Game {
-    fn event(&mut self, ctx: &mut EventCtx) {
-        self.app.per_obj.reset();
 
-        let transition = self.states.last_mut().unwrap().event(ctx, &mut self.app);
-        // If we fall through, there's a new state that we need to wakeup.
+    // If true, then the top-most state on the stack needs to be "woken up" with a fake mouseover
+    // event.
```
Comment (michaelkirk) [inline game/src/game.rs:82]: "fake mouseover" is this what the "no_op_event" is about?

Comment (dabreegster) [reply game/src/game.rs:82]: Yep. Probably could use a better name. I don't remember exactly the bug motivating this, but I think it was something like:
1) Click a button to enter some new state
2) Later, exit that state, but keep the mouse exactly in the same place
3) The button isn't in the hovering state until you generate a mouse event

File: game/src/info/mod.rs (hunk 2 above)
Comment (dabreegster) [reply game/src/info/mod.rs:535]: This was a little unrelated to this PR. It's a more ergonomic API for state constructors to return `Box<dyn State>`. But originally, I was making the caller do the boxing. I happened to clean this up for `SandboxMode` here, and this is the one weird consequence. When we rewind time, it's cleanest to just create a SandboxMode from scratch. The alternative to the rewrite here would be something like `sandbox = SandboxMode::new(ctx, app, sandbox.gameplay_mode).downcast::<SandboxMode>().ok().unwrap()`. That feels a little more confusing than this way

**Pull request merged**